swerve/SwerveModule.py

```diff
@@ -0,0 +1,214 @@
+
+    # create motors
+    # drive motor
+    self.drive_motor = hardware.TalonFX(drive_motor_id)
+    talonfx_configs = configs.TalonFXConfiguration()
+
+    # ----
+    self.drive_motor.configPeakOutputReverse(-RobotConstants.maxSpeed)
+    self.drive_motor.configPeakOutputForward(RobotConstants.maxSpeed)
+    self.drive_motor.configClosedLoopPeakOutput(0, RobotConstants.maxSpeed)
+    self.drive_motor.configFactoryDefault()
+    self.drive_motor.setNeutralMode(ctre.NeutralMode.Brake)
+    self.drive_motor.configStatorCurrentLimit(
+      ctre.StatorCurrentLimitConfiguration(True, 25, 30, 0.5)
+    )
+    self.drive_motor.configSupplyCurrentLimit(
+      ctre.SupplyCurrentLimitConfiguration(True, 30, 35, 0.5)
+    )
+    self.drive_motor.configOpenloopRamp(0.25)
+    self.drive_motor.setSelectedSensorPosition(0)
+    # self.drive_motor.configSelectedFeedbackCoefficient(
+    #     1 / SwerveConstants.kDrivingEncoderPositionFactor
+    # )
+    self.drive_motor.config_kP(0, 0)
+    self.drive_motor.config_kI(0, 0)
+    self.drive_motor.config_kD(0, 0)
+    # self.drive_motor.config_kF(0, 1)
+    self.drive_motor.config_kF(0, SwerveConstants.kFDriving)
+    # self.drive_motor.
+
+    # turn motor
+    self.turn_motor = rev.CANSparkMax(
+      turn_motor_id, rev.CANSparkMaxLowLevel.MotorType.kBrushless
+    )
+    # ?
+    self.turn_motor.restoreFactoryDefaults()
+    self.turn_motor.setInverted(turn_motor_reversed)
+    self.turn_motor.setIdleMode(rev.CANSparkMax.IdleMode.kBrake)
+    # self.turn_motor.setOpenLoopRampRate(200 / 1000)  # ms
+    self.turn_motor.setSmartCurrentLimit(15)
+    self.turn_motor.setSecondaryCurrentLimit(20)
+
+    # create encoders
+    self.turn_encoder = self.turn_motor.getAbsoluteEncoder(
+      rev.SparkMaxAbsoluteEncoder.Type.kDutyCycle
+    )
+    self.turn_encoder.setPositionConversionFactor(math.pi * 2)  # radians
+    self.turn_encoder.setVelocityConversionFactor(
+      (math.pi * 2) / 60
+    )  # radians per second
+    self.turn_encoder.setInverted(abs_encoder_reversed)
+    self.turn_encoder.setZeroOffset(abs_encoder_offset_rad)
+
+    # We are not using the built in PID controller, but this is where the config would go...
+    # self. #...
+    self.turn_pid = self.turn_motor.getPIDController()
+    self.turn_pid.setFeedbackDevice(self.turn_encoder)
+    self.turn_pid.setP(SwerveConstants.kPTurning)
+    self.turn_pid.setI(SwerveConstants.kITurning)
+    self.turn_pid.setD(SwerveConstants.kDTurning)
+    self.turn_pid.setPositionPIDWrappingEnabled(True)
+    self.turn_pid.setPositionPIDWrappingMinInput(0)
+    self.turn_pid.setPositionPIDWrappingMaxInput(math.pi * 2)
+    self.turn_pid.setOutputRange(-1, 1)
+
+    self.turn_motor.burnFlash()
+
+    self.resetEncoders()
+
+    self.initNetworkTables()
+
+    self.lastDesiresState = self.getState()
+
+  def initNetworkTables(self):
+    network_table_instance = NetworkTableInstance.getDefault()
+    self.swerve_dashboard = network_table_instance.getTable(
```

Comment:
Not sure if we'd do this over the network table.

If you use the shuffleboard, the data will be published to the network table and visually to the shuffleboard.